### ISSUE TITLE
tf2.4.0 and 2.4.1 keras history.history is different from each epoch end 

### ISSUE THREAD

**Issue by zhaoyun0071** [tensorflow/tensorflow#48033]
i run the code on colab
the code is :[https://colab.research.google.com/github/tensorflow/docs-l10n/blob/master/site/zh-cn/guide/keras/train_and_evaluate.ipynb?hl=zh_CN#scrollTo=mGYBLPAR13g3](url)

**and below is epcoh end** 

Fit model on training data
Epoch 1/2
782/782 [==============================] - 3s 3ms/step - loss: 0.5712 - sparse_categorical_accuracy: 0.8411 - val_loss: 0.2025 - val_sparse_categorical_accuracy: 0.9416
Epoch 2/2
782/782 [==============================] - 2s 3ms/step - loss: 0.1734 - sparse_categorical_accuracy: 0.9493 - val_loss: 0.1453 - val_sparse_categorical_accuracy: 0.9571

**and keras history.history below**

{'loss': [0.33999016880989075, 0.16297979652881622],
 'sparse_categorical_accuracy': [0.9027799963951111, 0.9526399970054626],
 'val_loss': [0.20245662331581116, 0.14533831179141998],
 'val_sparse_categorical_accuracy': [0.9416000247001648, 0.957099974155426]}

**i find the two is different.
for example train sparse_categorical_accuracy of epoch 1 is 0.8411 but in history.history is 0.9027799963951111.**

**but when i set verbose=2 in model.fit have no this problem**

**i try tf 2.3.2 and other low version have no question。**

**what changes in tf2.4.0 。i could not find the change log in ：https://github.com/tensorflow/tensorflow/blob/master/RELEASE.md**

**Comment by axeldavy** [comment]
Hi, I have the same issue, and I reported it in this related bug report:
https://github.com/tensorflow/tensorflow/issues/46713

**Comment by zhaoyun0071** [comment]
> Hi, I have the same issue, and I reported it in this related bug report:
> #46713

i think it may be not a bug，in 2.4.0， on epoch end 782/782  metrics is the 782th batch metrics。all train dataset metrics show in history.history
 2.3.2 and other low version on epoch end metrics is all train dataset metrics。

**Comment by Saduf2019** [comment]
@zhaoyun0071 
The code provided is fairly complex hence it would be difficult for us to pinpoint the issue. Could you please get the example down to the simplest possible repro? That will allow us to determine the source of the issue easily. Thanks!

**Comment by zhaoyun0071** [comment]
> @zhaoyun0071所提供的代码相当复杂，因此我们很难确定问题。
> 你能把例子归结为最简单的反驳吗？这将使我们能够很容易地确定问题的根源。谢谢！

`import tensorflow as tf
from tensorflow import keras
from tensorflow.keras import layers

inputs = keras.Input(shape=(784,), name="digits")
x = layers.Dense(64, activation="relu", name="dense_1")(inputs)
x = layers.Dense(64, activation="relu", name="dense_2")(x)
outputs = layers.Dense(10, activation="softmax", name="predictions")(x)

model = keras.Model(inputs=inputs, outputs=outputs)

(x_train, y_train), (x_test, y_test) = keras.datasets.mnist.load_data()

# Preprocess the data (these are NumPy arrays)
x_train = x_train.reshape(60000, 784).astype("float32") / 255
x_test = x_test.reshape(10000, 784).astype("float32") / 255

y_train = y_train.astype("float32")
y_test = y_test.astype("float32")

# Reserve 10,000 samples for validation
x_val = x_train[-10000:]
y_val = y_train[-10000:]
x_train = x_train[:-10000]
y_train = y_train[:-10000]


model.compile(
    optimizer=keras.optimizers.RMSprop(),  # Optimizer
    # Loss function to minimize
    loss=keras.losses.SparseCategoricalCrossentropy(),
    # List of metrics to monitor
    metrics=[keras.metrics.SparseCategoricalAccuracy()],
)


print("Fit model on training data")
history = model.fit(
    x_train,
    y_train,
    batch_size=64,
    epochs=1,
    # We pass some validation for
    # monitoring validation loss and metrics
    # at the end of each epoch
    validation_data=(x_val, y_val),
)


print(history.history)`

**Comment by Saduf2019** [comment]
@zhaoyun0071 
i ran the code shared on tf 2.4.1, tf 2.4.0 and nightly, i do not see any striking differences, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/64831a7dcb16b0eaf3e0c11b9fe40a8b/untitled580.ipynb)

**Comment by axeldavy** [comment]
If it is indeed the same issue than mine, last working should be tf 2.3, not tf 2.4.0. I tried the 2.5.0-rc0 branch a few days ago, and the issue is still there.

**Comment by zhaoyun0071** [comment]
> @zhaoyun0071
> i ran the code shared on tf 2.4.1, tf 2.4.0 and nightly, i do not see any striking differences, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/64831a7dcb16b0eaf3e0c11b9fe40a8b/untitled580.ipynb)

I mean 2.4.0 and 2.4.1have the same question。
they are different from 2.3.0 and other low version!
you can try 2.3.0,see the history loss and epoch loss

**Comment by zhaoyun0071** [comment]
> @zhaoyun0071
> i ran the code shared on tf 2.4.1, tf 2.4.0 and nightly, i do not see any striking differences, please find the [gist here](https://colab.research.google.com/gist/Saduf2019/64831a7dcb16b0eaf3e0c11b9fe40a8b/untitled580.ipynb)

you can try 2.3.0,see the history loss and epoch loss are same.
but 2.4.0  the history loss and epoch loss are different

**Comment by jvishnuvardhan** [comment]
I can reproduce the issue. The issue is mainly with `loss` and `sparse_categorical_accuracy` in training only. The validation metrics are matching with the  validation metrics in `history` object but the training metrics are not matching. [Here](https://colab.research.google.com/gist/jvishnuvardhan/83c64c997276e080d5fa11f59d4d33c3/untitled580.ipynb) is a gist for our reference. Thanks!

@zhaoyun0071 We will take a look at this issue and respond. Thanks for creating this issue.

**Comment by arcutright** [comment]
I also have this issue with the training metrics in tf-2.4.1 using python 3.8 and a custom loss function. It looks like the history object is correct and the debug info is incorrect (assuming the history values are calculated before updating the model weights, otherwise I have 3 different answers).

```
Epoch 20/20
6/6 [==============================] - 0s 78ms/step - loss: 0.0290
```
```python
history = model.fit(...)
# history['loss'][-1] = 0.04953611642122269
actual_losses = model.loss(y_true, y_pred).numpy()
# actual_losses[-1] = 0.050000716
```

**Comment by jvishnuvardhan** [comment]
@zhaoyun0071 Looks like this was resolved in recent `tf-nightly`. I couldn't reproduce the error anymore. Please check the [gist here](https://colab.research.google.com/gist/jvishnuvardhan/3aceaaeb392ce0baece7c1ab2d4d239f/untitled580.ipynb).

Please verify once and close the issue if this was resolved for you. Thanks!

**Comment by VladimirVisnovsky** [comment]
Hi, the problem persists in tensorflow/tensorflow:latest-gpu. By defining callback:
```
checkpoint_filepath = 'tmp/checkpoint'
model_checkpoint_callback = tf.keras.callbacks.ModelCheckpoint(
    filepath='tmp/6/model.{epoch:02d}-{AE loss min:.4f}',
    save_weights_only=True,
    monitor='AE loss min',
    mode='min',
    save_best_only=False,
    verbose=1,
    save_freq='epoch'
)
```
I get following fit output:

```
Epoch 1/5
159/163 [============================>.] - ETA: 0s - AE loss min: 0.3586 - AE loss max: 0.3586
Epoch 1: saving model to tmp/6/model.01-0.1878
163/163 [==============================] - 3s 8ms/step - AE loss min: 0.3534 - AE loss max: 0.3534
Epoch 2/5
162/163 [============================>.] - ETA: 0s - AE loss min: 0.1864 - AE loss max: 0.1864
Epoch 2: saving model to tmp/6/model.02-0.1659
163/163 [==============================] - 1s 9ms/step - AE loss min: 0.1861 - AE loss max: 0.1861
Epoch 3/5
156/163 [===========================>..] - ETA: 0s - AE loss min: 0.1633 - AE loss max: 0.1633 
Epoch 3: saving model to tmp/6/model.03-0.1576
163/163 [==============================] - 1s 6ms/step - AE loss min: 0.1631 - AE loss max: 0.1631
Epoch 4/5
162/163 [============================>.] - ETA: 0s - AE loss min: 0.1634 - AE loss max: 0.1634 
Epoch 4: saving model to tmp/6/model.04-0.1577
163/163 [==============================] - 1s 7ms/step - AE loss min: 0.1634 - AE loss max: 0.1634
Epoch 5/5
163/163 [==============================] - ETA: 0s - AE loss min: 0.1574 - AE loss max: 0.1574 
Epoch 5: saving model to tmp/6/model.05-0.1563
163/163 [==============================] - 1s 8ms/step - AE loss min: 0.1574 - AE loss max: 0.1574
```

And by calling model history I get:
```
{'AE loss min': [0.18779411911964417,
  0.1658637970685959,
  0.15759289264678955,
  0.15766046941280365,
  0.15625576674938202],
 'AE loss max': [0.18779411911964417,
  0.1658637970685959,
  0.15759289264678955,
  0.15766046941280365,
  0.15625576674938202]}
```
Expected output would be that every _AE loss min_ at the end of each epoch is equal across 
1. printed fit output 
2. printed ModelCheckpoint callback output
3. model history output

My TensorFlow version is 2.12.0 (checked on 2.11 with the same result). I am unable to install 2.3.0 due to conflicts in my environment and I think downgrade in this case is not a solution.
